lib/netext/grpcext/conn.go

```diff
@@ -0,0 +1,305 @@
+// Package grpcext allows gRPC requests collecting stats info.
+package grpcext
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+	"go.k6.io/k6/js/modules"
+	"go.k6.io/k6/metrics"
+
+	protov1 "github.com/golang/protobuf/proto" //nolint:staticcheck,nolintlint // this is the old v1 version
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	grpcstats "google.golang.org/grpc/stats"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/encoding/prototext"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/dynamicpb"
+)
+
+// Request represents a gRPC request.
+type Request struct {
+	MethodDescriptor protoreflect.MethodDescriptor
+	Tags             map[string]string
+	Message          []byte
+}
+
+// Response represents a gRPC response.
+type Response struct {
+	Message  interface{}
+	Error    interface{}
+	Headers  map[string][]string
+	Trailers map[string][]string
+	Status   codes.Code
+}
+
+type clientConnCloser interface {
+	grpc.ClientConnInterface
+	Close() error
+}
+
+// Conn is a gRPC client connection.
+type Conn struct {
+	raw clientConnCloser
+}
+
+// DefaultOptions generates an option set
+// with common options for requests from a VU.
+func DefaultOptions(vu modules.VU) []grpc.DialOption {
+	dialer := func(ctx context.Context, addr string) (net.Conn, error) {
+		return vu.State().Dialer.DialContext(ctx, "tcp", addr)
+	}
+
+	return []grpc.DialOption{
+		grpc.WithBlock(),
+		grpc.FailOnNonTempDialError(true),
+		grpc.WithReturnConnectionError(),
+		grpc.WithStatsHandler(statsHandler{vu: vu}),
+		grpc.WithContextDialer(dialer),
+	}
+}
+
+// Dial establish a gRPC connection.
+func Dial(ctx context.Context, addr string, options ...grpc.DialOption) (*Conn, error) {
+	conn, err := grpc.DialContext(ctx, addr, options...)
+	if err != nil {
+		return nil, err
+	}
+	return &Conn{
+		raw: conn,
+	}, nil
+}
+
+// ReflectionClient returns a reflection client based on the current connection.
+func (c *Conn) ReflectionClient() (*ReflectionClient, error) {
+	return &ReflectionClient{Conn: c.raw}, nil
+}
+
+// Invoke executes a unary gRPC request.
+//
+// TODO: should we support grpc.CallOption?
+func (c *Conn) Invoke(ctx context.Context, method string, md metadata.MD, req Request) (*Response, error) {
```

Comment:
> should we support grpc.CallOption?

Yes, I think this might be a good idea, given that they are present in [`grpc.ClientConn.Invoke()`](https://pkg.go.dev/google.golang.org/grpc#ClientConn.Invoke). Can you think of a reason why we shouldn't support them now, even if we don't use them yet?

Also, do we need a `method` parameter here when we have a `MethodDescriptor` in the request?

Reply:
Unfortunately, yes. The `FullName` of the method descriptor returns the full name domain as a dot notation. Instead, Invoke wants the `package.serviceName/methodName` format. I don't know if the simple replace of the last dot with a slash could be fine. Apparently, also the [protoc generator](https://github.com/golang/protobuf/blob/5d5e8c018a13017f9d5b8bf4fad64aaa42a87308/protoc-gen-go/grpc/grpc.go#L312) generates the url instead of use+replace the method descriptor's full name.